build-tools/scripts/PrepareWindows.targets

```diff
@@ -9,6 +9,10 @@
   <UsingTask AssemblyFile="$(_TopDir)\bin\Build$(Configuration)\xa-prep-tasks.dll" TaskName="Xamarin.Android.BuildTools.PrepTasks.JdkInfo" />
   <UsingTask AssemblyFile="$(_TopDir)\bin\Build$(Configuration)\xa-prep-tasks.dll" TaskName="Xamarin.Android.BuildTools.PrepTasks.ReplaceFileContents" />
   <Target Name="Prepare">
+    <Error
+        Text="The specified `%24(AndroidToolchainDirectory)` '$(AndroidToolchainDirectory)' contains a space. Android NDK commands do not support this. Please create a Configuration.Override.props file that sets the AndroidToolchainDirectory property to a different path."
+        Condition="$(AndroidToolchainDirectory.Contains (' '))"
+    />
```

Comment:
Can you attach a binlog of what is happening otherwise if you have a space in your username?

If you've not done it, you can run command line:
```
msbuild Xamarin.Android.sln /t:Prepare /bl
```
And a `msbuild.binlog` should show up in the current dir.

Reply:
Sure thing. The motivation for this change comes from the error in the build output that looks like this:

```
Project "C:\source\xamarin-android\Xamarin.Android.sln" (1) is building "C:\source\xamarin-android\src\sqlite-xamarin\sq
lite-xamarin.csproj" (44) on node 1 (default targets).
_NdkBuild:
  "C:\Users\Windows User\android-toolchain\ndk\ndk-build.cmd" SQLITE_SOURCE_DIR="C:\source\xamarin-android\external\sqlite" APP_ABI="armeabi-v7a x86"
  'C:\Users\Windows' is not recognized as an internal or external command,
  operable program or batch file.
C:\source\xamarin-android\src\sqlite-xamarin\sqlite-xamarin.targets(27,5): error MSB3073: The command ""C:\Users\Windows User\android-toolchain\ndk\ndk-build.cmd" SQLITE_SOURCE_DIR="C:\source\xamarin-android\external\sqlite" APP_ABI="armeabi-v7a x86"" exited with code 1. [C:\source\xamarin-android\src\sqlite-xamarin\sqlite-xamarin.csproj]
```

The full log is [msbuild-binlog.zip](https://github.com/xamarin/xamarin-android/files/2298776/msbuild-binlog.zip). It includes similar errors for `class-parse.exe` and `cmake`. The issues for `class-parse.exe` and `cmake` could be fixed by adjusting the quoting in the .targets files. But for `ndk-build.cmd`, the batch file *internally* references its own path (`%~dp0`) with no quotes, and that file comes directly from the NDK. So the build process would need to avoid `ndk-build.cmd` altogether to work when there are spaces in the path to the android-toolchain directory. For example, [src/Xamarin.Android.Build.Tasks/Tasks/Aot.cs](https://github.com/xamarin/xamarin-android/blob/32007a1d72df700b267ce61955f3ba60d9f2166d/src/Xamarin.Android.Build.Tasks/Tasks/Aot.cs#L378) and [src/Xamarin.Android.Build.Tasks/Tasks/MakeBundleNativeCodeExternal.cs](https://github.com/xamarin/xamarin-android/blob/32007a1d72df700b267ce61955f3ba60d9f2166d/src/Xamarin.Android.Build.Tasks/Tasks/MakeBundleNativeCodeExternal.cs#L143) use their own mechanism to locate binaries within the subdirectories of the NDK and run them directly rather than calling the batch files. That approach might also help protect against other characters like `%` that have special meanings in the context of `cmd.exe` batch files.